### ISSUE TITLE
Hang after publishing on a channel with confirms enabled

### ISSUE THREAD

**Issue by jbg** [sozu-proxy/lapin#245]
0.34.0 hangs after publishing on a channel with confirms enabled. Linux on arm64, compiled with latest rust nightly.

Out of curiosity - I've hit a few similar bugs with lapin that have caused hangs, and they always seem to block the entire runtime, like they're blocking the thread rather than just never polling `Ready`. This has got us looking at solutions for isolating lapin into its own runtime to make it possible to recover more gracefully from these bugs. I've not come across any similar bugs with any other library. Have I just been unlucky or is there something about AMQP that is difficult to implement within the contract of `Future`?

**Comment by jbg** [comment]
If I disable confirms on the channel the publish succeeds without hanging.

Publish code looks like:

```
tokio::time::timeout(
  TIMEOUT,
  channel.basic_publish(exchange_name, routing_key, publish_options, data.to_vec(), props)
).await??;

let nacked_messages = tokio::time::timeout(TIMEOUT, channel.wait_for_confirms()).await??;
if !nacked_messages.is_empty() {
  warn!("message nacked by RabbitMQ: {:?}", nacked_messages);
}
else {
  debug!("message acked by RabbitMQ");
}
```

Notably, the timeouts are ineffective because the runtime thread is blocked.

**Comment by Keruspe** [comment]
Which version of pinky-sweat are you using?
There was an issue priori to 1.3.1 where the waker wasn’t properly called
for this particular call using futures.

**Comment by jbg** [comment]
```
$ cargo metadata --format-version 1 | jq | grep pinky-swear | grep pkg
            "pkg": "pinky-swear 1.6.0 (registry+https://github.com/rust-lang/crates.io-index)",
```

**Comment by Keruspe** [comment]
Fwiw, the whole publisher-confirms code has been through a big rework as part of #243 (currenly nor in any release), and it might fix a bug when the server nacks the message, there are no other messages pending and we receive the basic-return before the basic-nack. This might (or might not) be your situation here?

Any chance to post trace logs around the hang?

**Comment by Keruspe** [comment]
WRT your other topic, which hangs have you seen with lapin in the past? Were those recent?
There definitely were issues at some point, but apart from two last points that need(ed) more attention, we've reached 1.0 stability in my opinion, so if you hit some of those recently I'd really be interested in some trace logs and context.

The two points being:
- the publisher confirms stuff that should hopefully land tomorrow as part of 0.35.0 alongside a couple of smaller changes
- #205 which will allow using the tokio or async-std or whichever other event loop instead of the default one shipped with lapin.

**Comment by Keruspe** [comment]
WRT AMQP and Future, yeah, AMQP is a really awkward protocol which is not really compatible with the "only run when polled" concept of Future, it's really more of a "push" protocol, which is the reason for using a custom io loop + pinky-swear (which is basically sugar around channels). This way the event loop can push data in the channel and notify subscribers (such a a Future's Context) once ready.
The `wait_for_confirm` call uses some exotic combinator from pinky-swear which had a bug *in its Future compat implementation* which didn't call the waker properly (`.wait()` worked but `.await` didn't "hang" per se, but the context was never woken so the future didn't get polled despite being ready (#244).

**Comment by jbg** [comment]
Trace logs around the hang:

```
[2020-04-01T05:47:19Z INFO  backends::modem_queues] Sending message to RabbitMQ...
[2020-04-01T05:47:19Z TRACE lapin::connection] connection send_frame; channel_id=1
[2020-04-01T05:47:19Z TRACE lapin::frames] channel 1 state is now waiting for ExpectedReply(ExchangeDeclareOk(Pinky))
[2020-04-01T05:47:19Z TRACE lapin::connection] connection wake
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:19Z TRACE pinky_swear] Called from future, registering waker
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-01T05:47:19Z TRACE lapin::io_loop] will write to buffer: Method(1, Exchange(Declare(Declare { exchange: ShortString("modems"), kind: ShortString("direct"), passive: false, durable: true, auto_delete: false, internal: false, nowait: false, arguments: FieldTable({}) })))
[2020-04-01T05:47:19Z TRACE pinky_swear] Resolving promise
[2020-04-01T05:47:19Z TRACE pinky_swear] Resolving promise gave: Err("SendError(..)")
[2020-04-01T05:47:19Z TRACE lapin::io_loop] wrote 38 bytes
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=false, status=Setup
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop run
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run; can_read=true, can_write=true, has_data=false
[2020-04-01T05:47:19Z TRACE lapin::io_loop] read 12 bytes
[2020-04-01T05:47:19Z TRACE lapin::connection] will handle frame: Method(1, Exchange(DeclareOk(DeclareOk)))
[2020-04-01T05:47:19Z TRACE pinky_swear] Resolving promise
[2020-04-01T05:47:19Z TRACE pinky_swear] Resolving promise gave: Ok(())
[2020-04-01T05:47:19Z TRACE pinky_swear] Got data, waking our waker
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop send continue; can_read=true, can_write=true, has_data=false
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run done; can_read=true, can_write=true, has_data=false, status=Setup
[2020-04-01T05:47:19Z TRACE pinky_swear] Called from future, registering waker
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop run
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:19Z DEBUG amqp] Declared exchange 'modems'
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run; can_read=true, can_write=true, has_data=false
[2020-04-01T05:47:19Z DEBUG amqp] Sending message to RabbitMQ...
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=false, status=Setup
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop run
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll
[2020-04-01T05:47:19Z TRACE lapin::connection] connection send_frames; channel_id=1
[2020-04-01T05:47:19Z TRACE lapin::connection] connection wake
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:19Z TRACE pinky_swear] Called from future, registering waker
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-01T05:47:19Z TRACE lapin::io_loop] will write to buffer: Method(1, Basic(Publish(Publish { exchange: ShortString("modems"), routing_key: ShortString("rack-12.frame-17.modem-104"), mandatory: true, immediate: false })))
[2020-04-01T05:47:19Z TRACE lapin::io_loop] wrote 41 bytes
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop send continue; can_read=false, can_write=true, has_data=true
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=true, status=Setup
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop run
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-01T05:47:19Z TRACE lapin::io_loop] will write to buffer: Header(1, 60, AMQPContentHeader { class_id: 60, weight: 0, body_size: 112, properties: AMQPProperties { content_type: None, content_encoding: Some(ShortString("msgpack")), headers: None, delivery_mode: Some(2), priority: None, correlation_id: None, reply_to: None, expiration: None, message_id: None, timestamp: None, kind: None, user_id: None, app_id: None, cluster_id: None } })
[2020-04-01T05:47:19Z TRACE lapin::io_loop] wrote 31 bytes
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop send continue; can_read=false, can_write=true, has_data=true
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=true, status=Setup
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop run
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run; can_read=false, can_write=true, has_data=true
[2020-04-01T05:47:19Z TRACE lapin::io_loop] will write to buffer: Body(1, [REDACTED])
[2020-04-01T05:47:19Z TRACE pinky_swear] Resolving promise
[2020-04-01T05:47:19Z TRACE pinky_swear] Resolving promise gave: Ok(())
[2020-04-01T05:47:19Z TRACE pinky_swear] Got data, waking our waker
[2020-04-01T05:47:19Z TRACE pinky_swear] Got data, notifying next in chain
[2020-04-01T05:47:19Z TRACE pinky_swear] Got data, waking our waker
[2020-04-01T05:47:19Z TRACE pinky_swear] Called from future, registering waker
[2020-04-01T05:47:19Z TRACE lapin::io_loop] wrote 120 bytes
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop do_run done; can_read=false, can_write=true, has_data=false, status=Setup
[2020-04-01T05:47:19Z TRACE lapin::channel] Waiting for pending confirms
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop run
[2020-04-01T05:47:19Z TRACE lapin::io_loop] io_loop poll
[2020-04-01T05:47:19Z TRACE pinky_swear] Called from future, registering waker
[2020-04-01T05:47:19Z TRACE pinky_swear] Called from future, registering waker
[2020-04-01T05:47:20Z TRACE lapin::io_loop] io_loop poll done
[2020-04-01T05:47:20Z TRACE lapin::io_loop] io_loop do_run; can_read=true, can_write=true, has_data=false
[2020-04-01T05:47:20Z TRACE lapin::io_loop] read 21 bytes
[2020-04-01T05:47:20Z TRACE lapin::connection] will handle frame: Method(1, Basic(Ack(Ack { delivery_tag: 1, multiple: false })))
[2020-04-01T05:47:20Z TRACE pinky_swear] Resolving promise
[2020-04-01T05:47:20Z TRACE pinky_swear] Resolving promise gave: Ok(())
[2020-04-01T05:47:20Z TRACE pinky_swear] Got data, notifying next in chain
```
Following the last log line, the runtime thread is hung indefinitely. Other futures don't get polled on it either.

Versions (I tried latest git master for this test, but the behaviour is the same on released 0.34.0):
```
pinky-swear 1.6.0 (registry+https://github.com/rust-lang/crates.io-index)
lapin 0.34.0 (git+https://github.com/sozu-proxy/lapin#9176ffa65040e5960567c838daf826e9f54b9b42)
```

**Comment by jbg** [comment]
> WRT your other topic, which hangs have you seen with lapin in the past? Were those recent?
> There definitely were issues at some point, but apart from two last points that need(ed) more attention, we've reached 1.0 stability in my opinion, so if you hit some of those recently I'd really be interested in some trace logs and context.

It's been a while since we've had a hang like this. The other ones I remember were something mid-last year where a very high message rate would cause some kind of infinite loop, and an issue we had in development where we'd get a hang upon trying to connect with TLS. The thing I found interesting about both cases was that the hang seemed to block the whole runtime thread (like `std::thread::sleep(infinite_duration)` would) rather than being a future that never resolves, but still lets the runtime do other work (like `tokio::time::delay_for(infinite_duration)` would). This makes the "blast radius" of the hang quite big; our system is designed to recover gracefully from most kinds of failure but endless blocking on one of the runtime threads is something that can only really be resolved by restarting the process.

> The two points being:
> 
>     * the publisher confirms stuff that should hopefully land tomorrow as part of 0.35.0 alongside a couple of smaller changes
>     * #205 which will allow using the tokio or async-std or whichever other event loop instead of the default one shipped with lapin.

#243 looks really useful, thanks! We would really love to be able to use the Tokio event loop too.

**Comment by Keruspe** [comment]
The good is that the two issues you mention are two of the three possible issues I was thinking of, the third one being much older. These were affecting the default io_loop shipped with lapin, which has received much more testing since then.

Wrt your particular issue, that's something else though.
There aren't any other logs after the one you pasted? No heartbeats or such?

**Comment by jbg** [comment]
With confirms enabled, there's nothing logged subsequently (I left it for several minutes).

With confirms disabled, I get the usual heartbeat noise regularly, after the publish succeeds.

**Comment by Keruspe** [comment]
Can you try with lapin 0.34.1 / pinky-swear 1.6.1 to see if any more logs get added?
I suspect not, but that could be a lead, if any

**Comment by Keruspe** [comment]
OH, I might have a clue. This *definitely* shouldn't block the main thread though (and will not once #205 gets further).
If I'm right, this is not an issue in lapin (well, not directly) but in pinky-swear. I'll definitely do lots more testing on this one, especially in the futures integration.

**Comment by jbg** [comment]
Logs seem to be the same with lapin 0.34.1 & pinky-swear 1.6.1

**Comment by Keruspe** [comment]
pinky-swear 1.7.0 could be better, but this will require further investigation anyways, even if it is better

**Comment by jbg** [comment]
The issue appears to be gone with pinky-swear 1.7.0, I'll confirm once the full suite of stress tests have run

**Comment by Keruspe** [comment]
Great ! Will track in #205 the fact that it blocked the *main* thread.

This gave me the opportunity to make a good cleanup inside of pinky-swear which will make it easier to audit and stress-test to make this sure this won't happen again in the future.

I also added this example https://github.com/sozu-proxy/lapin/commit/66b65779ac922b89df89d4681833597b5cd406a5 which I was about to ask you to run in the same env as your failing tests if 1.7.0 didn't fix your problem

**Comment by Keruspe** [comment]
FTR, the problem here was a dumb issue of upstream and downstream flows being under the same lock.

**Comment by Keruspe** [comment]
Ok, after giving it more thoughts, reading, testing, there were in fact two problems. The first one was the one stated above, the second one is that we were using a sync channel. I'm not sure why we used that in the first place. This is now fixed too in 0.35.1.

**Comment by jbg** [comment]
Actually, sorry, it could be when the channels are opened or when `confirm_select` is done - I'm digging a bit further to clarify that.

**Comment by Keruspe** [comment]
In the publisher_confirm example, this warning appears 8 times. 7 times it's the notification that a frame has been sent, 1 time something else

**Comment by Keruspe** [comment]
Ok, the other one is "normal".

When expecting a publisher confirm, we create a broadcaster to broadcast the completion to several places:
- the PublisherConfirm handle that is returned by basic_publish, which when dropped will cause this
- another internal place where we only store the last available confirmation promise for the "wait_for_confirms" call. The other ones being discarded on purpose, it's normal to hit this too.

WRT the 7 for frame sent notifications, I need to do some digging

**Comment by Keruspe** [comment]
Ok, so, for the publisher confirms one, I don't think they're really harmful, but we could do something to "deregister" the promise from the broadcaster to avoid that.

The other ones are clear now:
- when a method doesn't expect a response, the promise resolves when the frames are sent
- when a method expects a response, the promise from above is discarded and replaced by a promise which resolves when we get the response. We could chain in this case to avoid a warning, but the end result would basically be the same

**Comment by Keruspe** [comment]
Fixed the non publisher-confirms ones in 0.36.0

**Comment by Keruspe** [comment]
> When expecting a publisher confirm, we create a broadcaster to broadcast the completion to several places:
>     * the PublisherConfirm handle that is returned by basic_publish, which when dropped will cause this
>     * another internal place where we only store the last available confirmation promise for the "wait_for_confirms" call. The other ones being discarded on purpose, it's normal to hit this too.

The first part was wrong, when we drop the `PublisherConfirm`, we don't drop the associated promise, so the only case left is the second one, which is now tracked in #248 

I think we can close this issue now
